src/Ide/LocalBindings.hs

```diff
@@ -0,0 +1,197 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ide.LocalBindings
+  ( Bindings (..)
+  , bindings
+  , mostSpecificSpan
+  , holify
+  ) where
+
+import           Bag
+import           Control.Lens
+import           Data.Data.Lens
+import           Data.Function
+import           Data.Generics
+import           Data.List
+import           Data.Map (Map)
+import qualified Data.Map as M
+import           Data.Maybe
+import           Data.Ord
+import           Data.Set (Set)
+import qualified Data.Set as S
+import           GHC (TypecheckedModule (..), GhcTc, NoExt (..))
+import           HsBinds
+import           HsExpr
+import           Id
+import           OccName
+import           SrcLoc
+
+------------------------------------------------------------------------------
+-- | The available bindings at every point in a Haskell tree.
+data Bindings = Bindings
+  { bGlobalBinds :: Set Id
+  , bLocalBinds  :: Map SrcSpan (Set Id)
+  } deriving (Eq, Ord)
+
+instance Semigroup Bindings where
+  Bindings g1 l1 <> Bindings g2 l2 = Bindings (g1 <> g2) (l1 <> l2)
+
+instance Monoid Bindings where
+  mempty = Bindings mempty mempty
+
+
+------------------------------------------------------------------------------
+-- | Determine what bindings are in scope at every point in a program.
+--
+-- **WARNING:** This doesn't find bindings inside of TH splices or arrow syntax
+-- --- and possibly other obscure pieces of the AST.
+bindings :: TypecheckedModule -> Bindings
+bindings = uncurry Bindings . bindsBindings mempty . tm_typechecked_source
+
+
+------------------------------------------------------------------------------
+-- | Helper function for implementing 'binding'.
+--
+-- **WARNING:** This doesn't yet work over TH splices or arrow syntax --- and
+-- possibly other obscure pieces of the AST.
+dataBindings :: Data a => S.Set Id -> a -> M.Map SrcSpan (S.Set Id)
+dataBindings in_scope = foldMapOf biplate $ cool collect
```

Comment:
This information is computed in `.hie` files. If you look at the `refmap` in the result of the `GetHieFile` rule, you will get a map `Map Identifier [(Span, IdentifierDetails Type)]`. [`IdentifierDetails`](https://hackage.haskell.org/package/ghc-8.10.1/docs/HieTypes.html#t:IdentifierDetails) contains a set of `ContextInfo`, from which you recover the scopes for that particular identifier using [`getScopeFromContext`](https://hackage.haskell.org/package/ghc-8.10.1/docs/HieUtils.html#v:getScopeFromContext).

Reply:
I'd love to get rid of this; but if I'm reading your comment correctly, that mapping is the wrong direction. We want to go from a span to every binding in scope inside of it.

Reply:
Yes, you need to invert the map into something like a [`IntervalMap`](https://hackage.haskell.org/package/fingertree-0.1.4.2/docs/Data-IntervalMap-FingerTree.html), which should be pretty straightforward to construct. What we have is essentially a `Map Name [Scope]`, what we need is an `IntervalMap SrcLoc Name`, right?

Reply:
Delightfully good idea. I'll implement it today. Thanks!

Reply:
I gave this a go, and got as far as getting back a `TypeIndex` from hie. But I can't figure out how to reconstitute this into a proper GHC `Type`, which is necessary for how we've set up `refinery`. Holding off on moving forwards until someone can point me in the right direction on that one!

Reply:
See #396, that should make this work